### ISSUE TITLE
HowTo: Code sign agents directly on the Ubuntu server

### ISSUE THREAD

**Issue by jacotec** [Ylianst/MeshCentral#3571]
Hi,

as I've got our CodeSign cert today from Sectigo, I've looked for the ability to easily sign the meshagents after updates directly on the server. I've ended writing a short bash script which does this:

- MeshCentral is running in /opt/meshcentral
- The script copies the agents to /opt/Meshtemp
- Then the agents are signed with jsign
- Finally the signed versions are copied back and the service is restarted

The code signing cert is residing in /etc/ssl , I've exported it as "Sectigo.pfx".

My script is located in /opt, just run it with "./meshsign.sh".

I want to share this with the community, feel free to use / change or optimize it. Posting back any corrections here is appreciated.

jsign 4.0+ must be installed on the server: https://ebourg.github.io/jsign/

```
#!/bin/bash
 
if [ ! -d "/opt/Meshtemp" ]; then
  mkdir "/opt/Meshtemp"
fi
 
echo -e "Copying MeshService files to /opt/Meshtemp"
 
rm -f /opt/Meshtemp/MeshService* &&
cp /opt/meshcentral/node_modules/meshcentral/agents/MeshService.exe /opt/Meshtemp/ &&
cp /opt/meshcentral/node_modules/meshcentral/agents/MeshService64.exe /opt/Meshtemp/ &&
 
echo -e "Signing files"
 
jsign --keystore /etc/ssl/Sectigo.pfx --keypass ***keypassword*** --tsaurl http://timestamp.sectigo.com --tsmode RFC3161 --replace /opt/Meshtemp/MeshService.exe &&
jsign --keystore /etc/ssl/Sectigo.pfx --keypass ***keypassword*** --tsaurl http://timestamp.sectigo.com --tsmode RFC3161 --replace /opt/Meshtemp/MeshService64.exe &&
 
echo -e "Moving signed files back"
 
mv /opt/Meshtemp/MeshService.exe /opt/Meshtemp/MeshService-signed.exe &&
mv /opt/Meshtemp/MeshService64.exe /opt/Meshtemp/MeshService64-signed.exe &&
cp /opt/Meshtemp/MeshService-signed.exe /opt/meshcentral/node_modules/meshcentral/agents/MeshService-signed.exe &&
cp /opt/Meshtemp/MeshService64-signed.exe /opt/meshcentral/node_modules/meshcentral/agents/MeshService64-signed.exe &&
 
echo -e "Restarting MeshCentral"
 
service meshcentral restart &&
 
echo -e "Task successfully completed."
```

**Comment by JSuenram** [comment]
Sure works, but you won't get SmartScreen-Reputation and will still have warning.
Signing for Windows 10 means having the private key ob Hardware USB Thumb or in Azure Key Chain.
Only then you will get Smartscreen-Reputation and warning messages from Windows will reduce.

**Comment by jacotec** [comment]
@JSuenram So you're talking about an EV cert?

They are horrible expensive. "Normal" ones should also work, but may take a bit time to build reputation (see: https://stackoverflow.com/a/66582477 )

EUR 76.- / year vs. $400+ / year makes a difference for private people or smaller companies.

**Comment by JSuenram** [comment]
> @JSuenram So you're talking about an EV cert?
> 
> They are horrible expensive. "Normal" ones should also work, but may take a bit time to build reputation (see: https://stackoverflow.com/a/66582477 )
> 
> EUR 76.- / year vs. $400+ / year makes a difference for private people or smaller companies.

Yeah, absolutely right.... they are to expensive and you pay for some sort of "VIP"... the validation is useless.
And to be clear; All this Cert-Shit is only and just cause Windows is so insecure by design....

But it works like that and SmartScreen never complained again, 5 seconds after signing with CodeSign....

**Comment by jacotec** [comment]
@Ylianst I'm excited about the new code signing feature ;-)

Thanks for mentioning the meshcentral-data/agents thing. I'll change it to that then. This is great!

Question/Feature request: It would be very helpful if we could get some notification from the server when the agent was updated (i.e. an email to admin), so we can take action to re-sign and update. It saves the time looking for changed agents after each update of the MC server (as you told: Agents are much less frequently updated)

**Comment by jacotec** [comment]
@Ylianst Question: In meshcentral-data/agents only the two files

- MeshService-signed.exe
- MeshService64-signed.exe

are needed? Not the files without "-signed" from the subfolder in "node-modules"?

**Comment by Ylianst** [comment]
Right, only these two files. MeshCentral will grab the `-signed` files first. If there are none, it will use the unsigned versions.

**Comment by JSuenram** [comment]
@jacotec 
If your question is answered. Pls close....
